reference-implementation/lib/readable-byte-stream.js

```diff
@@ -0,0 +1,965 @@
+}
+
+function EnqueueInReadableByteStreamController(controller, chunk) {
+  controller._queue.push({buffer: chunk.buffer, byteOffset: chunk.byteOffset, byteLength: chunk.byteLength});
+  controller._totalQueuedBytes += chunk.byteLength;
+}
+
+function ErrorReadableByteStream(stream, e) {
+  assert(IsReadableByteStream(stream), 'stream must be ReadableByteStream');
+  assert(stream._state === 'readable', 'state must be readable');
+
+  stream._state = 'errored';
+  stream._storedError = e;
+
+  if (IsReadableByteStreamLocked(stream)) {
+    ErrorReadableByteStreamReader(stream._reader, e)
+  }
+}
+
+function ErrorReadableByteStreamReader(reader, e) {
+  reader._state = 'errored';
+  reader._storedError = e;
+
+  reader._closedPromise_reject(e);
+  reader._closedPromise_resolve = undefined;
+  reader._closedPromise_reject = undefined;
+
+  if (IsReadableByteStreamReader(reader)) {
+    for (const req of reader._readRequests) {
+      req.reject(reader._storedError);
+    }
+
+    reader._readRequests = [];
+  } else {
+    assert(IsReadableByteStreamByobReader(reader), 'reader must be ReadableByteStreamByobReader');
+
+    for (const req of reader._readIntoRequests) {
+      req.reject(reader._storedError);
+    }
+
+    reader._readIntoRequests = [];
+  }
+
+  DetachReadableByteStreamReader(reader);
+}
+
+function FillPullIntoDescriptorFromQueue(controller, pullIntoDescriptor) {
+  const elementSize = pullIntoDescriptor.elementSize;
+
+  const currentAlignedBytes = pullIntoDescriptor.bytesFilled - pullIntoDescriptor.bytesFilled % elementSize;
+
+  const maxBytesToCopy = Math.min(controller._totalQueuedBytes,
+                                  pullIntoDescriptor.byteLength - pullIntoDescriptor.bytesFilled);
+  const maxBytesFilled = pullIntoDescriptor.bytesFilled + maxBytesToCopy;
+  const maxAlignedBytes = maxBytesFilled - maxBytesFilled % elementSize;
+
+  let totalBytesToCopyRemaining = maxBytesToCopy;
+  let ready = false;
+  if (maxAlignedBytes > currentAlignedBytes) {
+    totalBytesToCopyRemaining = maxAlignedBytes - pullIntoDescriptor.bytesFilled;
+    ready = true;
+  }
+
+  const queue = controller._queue;
+
+  while (totalBytesToCopyRemaining > 0) {
+    const headOfQueue = queue[0];
+
+    const bytesToCopy = Math.min(totalBytesToCopyRemaining, headOfQueue.byteLength);
+
+    const destStart = pullIntoDescriptor.byteOffset + pullIntoDescriptor.bytesFilled;
+    new Uint8Array(pullIntoDescriptor.buffer).set(
+        new Uint8Array(headOfQueue.buffer, headOfQueue.byteOffset, bytesToCopy), destStart);
+
+    if (headOfQueue.byteLength === bytesToCopy) {
+      queue.shift();
+    } else {
+      headOfQueue.byteOffset += bytesToCopy;
+      headOfQueue.byteLength -= bytesToCopy;
+    }
+    controller._totalQueuedBytes -= bytesToCopy;
+
+    pullIntoDescriptor.bytesFilled += bytesToCopy;
+
+    totalBytesToCopyRemaining -= bytesToCopy
+  }
+
+  return ready;
+}
+
+function InitializeReadableByteStreamReader(reader, stream) {
```

Comment:
Again the function names are a bit confusing. Maybe name the ones that are shared e.g. InitializeReadableByteStreamReaderShared or something?